### DISCUSSION TITLE
How to get PlutoGrid value from PlutoGridStateManager

### DISCUSSION THREAD

**Discussion by allenlim-0627** [bosskmk/pluto_grid#129] [Q&A]
Hi, I am currently trying to get values from PlutoGrid such as text type column value or selection type column value, may I know how do I do that? Can I get all the inserted value from PlutoGridStateManager and convert to a class?

Thank you.

**Comment by bosskmk** [comment]
`stateManager.rows`
> You can access rows in List<PlutoRow> type.
> However, if they are filtered, you can only access them.


`stateManager.refRows.originalList`
> Same as rows, but you can access all rows regardless of filtering or not.


Every row has a cells property and this is a map.
The field value specified in the column can be accessed with the key.

Suppose you have a field called user_id.
If you need to get the user_id values of all rows into a List, you can do as follows.

```dart
var user_ids = stateManager.refRows.originalList.map((e) => e.cells['user_id'].value.toString());
```

If user_id is retrieved only from the currently displayed rows in the filtered state, it is as follows.

```dart
var user_ids = stateManager.rows.map((e) => e.cells['user_id'].value.toString());
var user_ids = stateManager.refRows.map((e) => e.cells['user_id'].value.toString());
```

rows and refRows are the same. However, rows are just destructured.

**Reply by allenlim-0627** [reply]
Thank you for the answer, I've tried it and it works. May I ask another question regarding the Selection type column? How do I reset the value to null or empty after I choose the first element of the drop-down list. Suppose I have a selection type column called user_type and the first element of the selection list is "Add New User", how can I reset the selection value after I click the "Add New User"?

I've tried to use stateManager.clearCurrentCell() or stateManager.setCurrentCell(PlutoCell(value: null, rowInx)) in the onChanged event of the PlutoGrid widget but no luck.

Thank you.

**Reply by bosskmk** [reply]
In the case of the selection column, values other than items are not changed.

Currently, you can change the value in the following way.

```dart
stateManager.currentCell.value = '';
stateManager.notifyListeners();
```

I will add a method to reset the value later.


And changing the value is as follows.

```dart
// For force, the default is false. 
// This is when the mode of the grid is select, 
// only the row can be selected and the value cannot be modified. 
// In this case, the value can be forcibly modified.
stateManager.changeCellValue(stateManager.currentCell.key, null, force: true);
```

**Reply by allenlim-0627** [reply]
Hi bossmk, once again thank you for the answer, may I know if the method to reset the value is already implemented in the newer version? I have another question to ask regarding the enableRowChecked properties in PlutoColumn. Is there any way for me to implement a custom Function once the checkbox of a row is checked?

Thank you.

**Reply by bosskmk** [reply]
reset the value hasn't worked yet.

The processing according to whether row is check or uncheck is as follows.

### When creating a grid, register a callback function with addListener in the stateManager.
```dart
      body: PlutoGrid(
        columns: columns,
        rows: rows,
        onChanged: (PlutoGridOnChangedEvent event) {
          print(event);
        },
        onLoaded: (PlutoGridOnLoadedEvent event) {
          event.stateManager.setSelectingMode(PlutoGridSelectingMode.row);

          stateManager = event.stateManager;

          stateManager.addListener(handleCheckedRow);
        },
        // configuration: PlutoConfiguration.dark(),
      ),
```

### Access checked or unchecked rows with `stateManager.checkedRows` or `stateManager.unCheckedRows` property in callback function.
```dart
  List<PlutoRow> checkedRows = [];

  void handleCheckedRow() {
    var currentCheckedRows = stateManager.checkedRows;

    if (!listEquals(checkedRows, currentCheckedRows)) {
      checkedRows = stateManager.checkedRows;
      print('changed!');
    } else {
      print('not changed!');
    }
  }
```

> Note: In the above example, even if the selected rows of the listEquals function are the same, even if the position in the list is changed by drag, the changed! Works as.
